ledger-api/grpc-definitions/com/digitalasset/ledger/api/v1/command_completion_service.proto

```diff
@@ -62,28 +61,50 @@ message CompletionStreamRequest {
   LedgerOffset offset = 4;
 }
 
+// completion stream responses always start with the snapshot of
+// active set of commands and subsequently only updates to that ACS
 message CompletionStreamResponse {
 
-  // This checkpoint may be used to restart consumption.  The
-  // checkpoint is after any completions in this response.
-  // Optional
-  Checkpoint checkpoint = 1;
+  message Update {
+
+    // This checkpoint may be used to restart consumption.  The
+    // checkpoint is after any completions in this response.
+    // Optional
+    Checkpoint checkpoint = 1;
+
+    // If set, one or more completions.
+    repeated Completion completions = 2;
 
-  // If set, one or more completions.
-  repeated Completion completions = 2;
+    // If set, one or more status updates
+    repeated StatusUpdate statusUpdates = 3;
+  }
+
+  message Snapshot {
+    Checkpoint checkpoint = 1;
+    repeated StatusUpdate activeCommands = 1;
+  }
+
+  oneof response {
+    Snapshot = 1; // first response will always be snapshot
+    Update = 2; // subsequent will be updates
+  }
 
 }
 
+message StatusUpdate {
+  string command_id = 1; // user supplied command id
+  string transaction_id = 2; // internally generated transaction_id off command_id
+  enum Status {
+    QUEUED = 1; // command has been locally received by participant and queued for submission
```

Comment:
Such interface would have profound effect on participant implementation. Being able to respond to client with a status of his request in a reliable manner while at the same time satisfying HA and HS requirements implies a database or a queuing system in the submission pipeline. It is something we were trying to avoid in the original design as it has inevitable performance ramifications.